### DISCUSSION TITLE
Hiding subprograms from menu and search

### DISCUSSION THREAD

**Discussion by harisreedhar** [JacquesLucke/animation_nodes#1710] [Ideas]
Currently animation nodes shows every subprograms in the menu and search. Which makes menu cluttered and less accessible. 
It would be nice to have an option to hide. I have a simple solution for it by hiding the name starts with an `"__"` and making it only accessible through invoke subprogram node.

![Screenshot from 2021-05-02 23-36-58](https://user-images.githubusercontent.com/46858047/116823007-969db600-ab9f-11eb-9790-2c0ca1218170.png)

**Comment by OmarEmaraDev** [comment]
What if we put subprograms in a sub-menu instead?

**Reply by Gerstmann-Bradley** [reply]
Would you elaborate more about sub-menu?
I agree with hidden as well. Frequently I create subprogram to speed up the workflow. There are some subprograms I want to use as freely as built-in nodes (I meant for the search function), but there also other nodes that I may not want to see at all.

**Reply by OmarEmaraDev** [reply]
Perhaps I misunderstood the intend. Where does the clutter happen? Why does the existence of many subprograms produce clutter? And which subprograms do we hide and which do we show?

**Reply by Gerstmann-Bradley** [reply]
Sorry, I and Hari may not talk the same. But he mentioned "menu" and _"search"_ in the title, so...
I personally only concern about "search". There's no real clutter due to search filter, but there may still be some nodes, which I don't want to see. For example, a smaller subprogram is used to build up a larger subprogram, while the smaller subprogram by itself is not necessarily useful. Smaller subprogram may be used for various purpose, such as help vectorization... I usually just name them casually to prevent I find out these nodes in the "search", but this also caused an issue in naming.

As for "which subprogram do we hide and which do we show", I agree with Hari's proposal if possible.
He mentioned that based on certain marker in the name, such as "_", the subprogram will be hidden. We may not need to use the same marker as he suggested, but I think it's a very good idea to differentiate based on the marker in the name.

As for the menu part that Hari demonstrated in his figure, I am not very interested in since I have never used menu.

At the end, since I don't really have clutter in "search". I'm neutral with this proposal. It will be nice if this can be done, but if this is complicate to do then I don't mind to prioritize other things first.

**Reply by OmarEmaraDev** [reply]
I see what you mean. Maybe we can approach this differently though.
Maybe we can add an option in the advanced node settings of the subprograms to let them define which category they want to appear it. Category menus would then add a sub-menu listing all subprograms belonging to that category. If the subprogram doesn't belong to a category, it is added in a sub-menu in the subprograms category to prevent clutter.

Not sure about search though. It is assumed that search is global and you will be able to find everything. So omitting things may not be the best idea. But maybe we can approach this differently as well somehow.

**Reply by Gerstmann-Bradley** [reply]
I am ok with this.

**Reply by harisreedhar** [reply]
By clutter i actually mean it so confusing when we have so many nested subprograms. All the inner subprograms will appear on search as well as on the menu. which is not always useful.
Creating sub-menu is good idea but sometimes we purposefully hide certain subprograms. for example when we share a node with someone it should be clean and show only that one node, not all subprograms inside it.


https://user-images.githubusercontent.com/46858047/116827416-7e856100-abb6-11eb-9c69-38c807c653dc.mp4

**Reply by harisreedhar** [reply]
May be we can add an option in settings to show hidden nodes as well.

**Reply by OmarEmaraDev** [reply]
@harisreedhar But I think hiding it everywhere except the Invoke node is a bit aggressive. I never added an invoke node in my life, and I guess that applies to most users. So as far as the user is concerned, those subprograms are lost. This needs to be handled differently.

**Reply by harisreedhar** [reply]
What about node tree specific visibility?

**Reply by OmarEmaraDev** [reply]
So each subprogram declare which node trees it wants to be visible in? What if those were imported to another file?

**Reply by 3DSinghVFX** [reply]
> I see what you mean. Maybe we can approach this differently though.
> Maybe we can add an option in the advanced node settings of the subprograms to let them define which category they want to appear it. Category menus would then add a sub-menu listing all subprograms belonging to that category. If the subprogram doesn't belong to a category, it is added in a sub-menu in the subprograms category to prevent clutter.

@OmarEmaraDev This is a good idea. Is it possible to define/name/remove the category at any point?
